# Pandas for AI/ML – Conceptual & Technical Assessment


### Aim
To assess students' knowledge and application of Pandas concepts and their ability to perform data manipulation and analysis using Python.
### Learning Outcomes
By the end of the assignment, students will be able to:
1. Interpret Pandas code snippets and explain the concepts involved in data manipulation, analysis, and cleaning.
2. Identify and diagnose errors in Pandas code and apply appropriate debugging techniques.
3. Apply Pandas techniques for data manipulation, analysis, cleaning, and handling edge cases in real-world datasets.
4. Evaluate real-world scenarios and justify appropriate Pandas-based solutions for data manipulation, analysis, and cleaning.

---

## Section A – Output Prediction

Predict the exact output of the given code snippets and explain the underlying Pandas and NumPy operations.

### Q1. Array Operations and Vectorization — 2 Marks

**Question:**

1. Predict the output of both statements.

**Predicted Output:**

```text
[20 40 60]

0    20
1    40
2    60
dtype: int64
```
2. Identify one similarity and one difference between the two data structures.

**Similarity:** Both NumPy arrays and Pandas Series perform element-wise operations.

**Difference:** A NumPy array does not have a labeled index by default, while a Pandas Series has an index for its values.th their index labels.


In [96]:
import numpy as np                     # Import the NumPy library
import pandas as pd                    # Import the Pandas library

arr = np.array([10, 20, 30])           # Create a NumPy array

s = pd.Series([10, 20, 30])            # Create a Pandas Series

print(arr * 2)                         # Multiply each array element by 2 and print the result

print(s * 2)                           # Multiply each Series value by 2 and print the result

[20 40 60]
0    20
1    40
2    60
dtype: int64


### Explanation

`np.array()` and `pd.Series()` store data differently, but `* 2` multiplies every element by 2 in both.

---

### Q2. Boolean Filtering

1. Predict the rows displayed.

**Answer:**

**Predicted Output:**
```text
  Student  Marks
1       B     80
2       C     65
3       D     90
```
Number of Rows: **3**

Number of Columns: **2**

2. How many rows and columns will the resulting DataFrame contain?

**Answer:**  
The resulting DataFrame contains 3 students and 2 columns: `Student` and `Marks`.

Number of Rows: **3**

Number of Columns: **2**

In [97]:
import pandas as pd                         # Import the Pandas library

df = pd.DataFrame({                         # Create a DataFrame
    'Student': ['A', 'B', 'C', 'D'],        # Create the Student column
    'Marks': [45, 80, 65, 90]               # Create the Marks column
})

result = df[df["Marks"] >= 65]              # Select rows where Marks are 65 or more

print(result[["Student", "Marks"]])          # Display only Student and Marks columns

  Student  Marks
1       B     80
2       C     65
3       D     90


**Explanation:**  
The condition `df['Marks'] >= 65` selects only the rows where the Marks value is 65 or greater. Therefore, students B, C, and D are selected.

---





### Q3. Descriptive Statistics — 2 Marks

**Questions:**

1. Predict all three outputs.

**Predicted Output:**

```text
30.0
50
30.0
```
2. Explain how the median differs from the mean.

Mean: The mean is the average of all the values. It is calculated by adding all values and dividing by the number of values.

Median: The median is the middle value when the data is arranged in ascending order.
```text
For example, in [10, 20, 30, 40, 50]:

Mean = (10 + 20 + 30 + 40 + 50) / 5 = 30
Median = middle value = 30
```

In [98]:
# Q3. Descriptive Statistics

import pandas as pd  # Import the Pandas library

df = pd.DataFrame({                # Create a DataFrame
    'Marks': [10, 20, 30, 40, 50]  # Create a column named 'Marks' with five values
})

print(df['Marks'].mean())          # Calculate and print the mean of the Marks column

print(df['Marks'].max())           # Find and print the maximum value in the Marks column

print(df['Marks'].median())        # Calculate and print the median of the Marks column

30.0
50
30.0


**Explanation:**

The mean is the average of all the values, while the median is the middle value when the values are arranged in ascending order.

For the given data `[10, 20, 30, 40, 50]`, both the mean and median are `30.0`.

---

### Q4. Group-wise Analysis

1. Predict the output.

**Predicted Output:**

```text
Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64
```


2. Explain how groupby() calculates the department-wise average.

**Ans:** The `groupby()` function groups the data based on the `Department` column and then calculates the average salary for each department.

- HR average = `(30000 + 50000) / 2 = 40000`
- IT average = `(40000 + 60000) / 2 = 50000`

Therefore, the department-wise average salaries are **HR: 40000.0** and **IT: 50000.0**.

In [99]:
import pandas as pd                         # Import the Pandas library

df = pd.DataFrame({                         # Create a DataFrame
    'Department': ['IT', 'HR', 'IT', 'HR'], # Create the Department column
    'Salary': [40000, 30000, 60000, 50000]  # Create the Salary column
})

print(df.groupby('Department')['Salary'].mean())  # Group by Department and calculate average Salary

Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64


### Explanation

`df.groupby('Department')` groups the rows based on the values in the `Department` column.
`['Salary']` selects the `Salary` column from each department group.
`.mean()` calculates the average salary for each department group.
Therefore, the code groups the employees by department and calculates the average salary separately for each department.

---

### Q5. Pandas Correlation - Positive

1. Predict the correlation coefficient.

**Predicted Output:**
`1.0`

2. Interpret the direction and strength of the relationship.

**Interpretation:**

The relationship is **positive and very strong (perfect positive correlation)**. As the number of Hours increases, the Marks also increase consistently.



In [100]:
import pandas as pd                         # Import the Pandas library

df = pd.DataFrame({                         # Create a DataFrame
    'Hours': [1, 2, 3, 4],                  # Create the Hours column
    'Marks': [20, 40, 60, 80]               # Create the Marks column
})

print(df['Hours'].corr(df['Marks']))        # Calculate and print the correlation coefficient

1.0


**Explanation**

`.corr()` selects the two specified columns and calculates their linear correlation coefficient.
It measures how strongly the `Hours` and `Marks` values are linearly related.

---

### Q6. Pandas Correlation - Negative

1. Predict the correlation coefficient.

**Predicted Output:**
`-1.0`


2. Explain what the sign of the coefficient indicates about the relationship.

`.corr()` calculates the linear correlation coefficient between the `Price` and `Demand` columns.
The result shows the direction and strength of their linear relationship.


In [101]:
import pandas as pd                         # Import the Pandas library

df = pd.DataFrame({                         # Create a DataFrame
    'Price': [10, 20, 30, 40],              # Create the Price column
    'Demand': [100, 80, 60, 40]             # Create the Demand column
})

print(df['Price'].corr(df['Demand']))       # Calculate and print the correlation coefficient

-1.0


### Explanation

The code creates a DataFrame with `Price` and `Demand` columns and uses `.corr()` to calculate their linear correlation coefficient.

---
---

# Section B – Code Correction
Covers the interpretation of Pandas code. Assesses logical reasoning and the ability to predict code outputs without execution. 

### Q1. JSON and CSV

**Question:**
1. A CSV file contains a header row, but the following code incorrectly reads the first student's record as column names. Correct the code.

**Issue:**

`header=None` tells Pandas that the CSV file does not contain a header row.

**Corrected Code:**

Use `header=0` so Pandas reads the first row of the CSV file as the column names.

In [102]:
import pandas as pd                         # Import the Pandas library

# CHANGED: header=None to header=0 tells Pandas that the first row contains column names
df = pd.read_csv(r"C:\Users\Lenovo\Downloads\students.csv", header=0)

print(df)                                   # Display the DataFrame

    ID     Name        Date  Marks
0  101  Asha, R  2026-09-01   80.0
1  102     Ravi  01/09/2026    NaN
2  103     Neha  2026-09-03    NaN


**Explanation**

`pd.read_csv()` reads the CSV file into a DataFrame, and `header=0` uses the first row as the column names.

---

### Q2. Data Analysis

**Question:**

2. The objective is to display the five highest salaries from a DataFrame. Correct the code.

**Issue:**

`sort_values('Salary')` sorts the salaries in ascending order by default, so `head(5)` returns the five lowest salaries.

**Correction:**

Sort the `Salary` column in descending order using `ascending=False`, then use `head(5)`.

In [103]:
import pandas as pd                              # Import the Pandas library

df = pd.DataFrame({                              # Create a DataFrame
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'], # Create the Employee column
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000] # Create the Salary column
})
# CHANGED: ascending=False sorts salaries from highest to lowest
print(df.sort_values('Salary', ascending=False).head(5)) # Sort salaries from highest to lowest and display top 5

  Employee  Salary
5        F   70000
3        D   60000
1        B   50000
4        E   45000
0        A   30000


### Explanation

`sort_values()` sorts the DataFrame based on the `Salary` column, and `ascending=False` sorts it from highest to lowest.
`head(5)` then selects the first five rows.

---

### Q3. Cleaning Data

**Question:**

3. A student wants to replace missing marks with the column's mean but writes code that does not update the DataFrame. Correct the code.

**Issue:**

`fillna()` returns the updated values, but the result was not assigned back to the `Marks` column.

**Correction:**

Assign the result of `fillna()` back to `df['Marks']` to update the DataFrame.

In [104]:
import pandas as pd                         # Import the Pandas library

df = pd.DataFrame({'Marks': [70, None, 90]}) # Create a DataFrame with a missing value

# CHANGED: Assigned the fillna() result back to Marks to update the DataFrame
df['Marks'] = df['Marks'].fillna(df['Marks'].mean()) # Replace the missing value with the mean and update the column

print(df)                                   # Display the updated DataFrame

   Marks
0   70.0
1   80.0
2   90.0


**Explanation**

`fillna()` replaces the missing value with the column mean, and assigning it back to `df['Marks']` updates the DataFrame.

---

### Q4. Correlation

**Question:**

4. The objective is to calculate the correlation between `Hours` and `Marks`. Correct the code.

**Issue:**

`df.corr()` does not take column names as separate arguments like `df.corr('Hours', 'Marks')`.

**Correction:**

Select the two columns first and then use `.corr()` to calculate their correlation.

In [105]:
import pandas as pd                         # Import the Pandas library

df = pd.DataFrame({                         # Create a DataFrame
    'Hours': [1, 2, 3, 4],                  # Create the Hours column
    'Marks': [20, 40, 60, 80]               # Create the Marks column
})
# CHANGED: Selected Hours and Marks first, then used corr()
print(df[['Hours', 'Marks']].corr())        # Select both columns and calculate their correlation

       Hours  Marks
Hours    1.0    1.0
Marks    1.0    1.0


**Explanation**

`df[['Hours', 'Marks']]` selects the two columns, and `.corr()` calculates the correlation between them.

---
---

# Section C – Edge Cases
Covers the practical implementation of Pandas operations. 
Assesses students' ability to write, execute, and explain Pandas code for numerical computations, preprocessing, and AI/ML tasks. 





## Q1. NumPy Broadcasting
**Question:**

1. Two NumPy arrays have shapes `(3, 1)` and `(1, 3)`. Predict the shape of their element-wise sum and explain how broadcasting works. Why might the result differ from adding two one-dimensional arrays?

**Ans:** The element-wise sum of arrays with shapes `(3, 1)` and `(1, 3)` will have the shape **`(3, 3)`**.

NumPy broadcasting allows arrays with compatible but different shapes to be used in arithmetic operations. Here:

- Array `a` has shape `(3, 1)`.
- Array `b` has shape `(1, 3)`.
- NumPy broadcasts both arrays to shape `(3, 3)` and performs element-wise addition.

Given:

```python
import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

result = a + b

print(result)
print(result.shape)
```

**1a. Predict the output of the given code.**

**Ans. predicted Output:**
```
[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)
```
**1b. Identify the shapes of arrays `a` and `b` before addition.**

**Ans. Shapes of arrays `a` and `b` before addition:**
- Shape of `a` = (3, 1)
- Shape of `b` = (1, 3)

**1c.Explain how NumPy broadcasting enables the addition of these arrays.**

**Ans: Broadcasting:**

NumPy broadcasting allows arrays with compatible shapes to perform element-wise operations.

Here, `a` has shape (3, 1) and `b` has shape (1, 3). NumPy broadcasts them to a compatible shape of (3, 3) and performs the addition.


**1d. What would happen if `b` were replaced with `np.array([1, 2])`? Explain the reason.**

**Ans. If `b = np.array([1, 2])`:**

The operation will raise a `ValueError` because the shapes (3, 1) and (2,) are not compatible for broadcasting.

Therefore, the arrays cannot be added element-wise.

In [106]:
import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

result = a + b         #adds the two arrays together, broadcasting the single column of a across the three columns of b

print(result)
print(result.shape)    #displays the shape of the resulting array, which is (3, 3)

[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)


---

## Q2. CSV Parsing, Missing Values and Date Formats

### Question

2. A CSV file contains commas inside quoted text fields, missing values, and dates in different formats. Explain how incorrect parsing can affect the resulting DataFrame and identify suitable read options or preprocessing steps.

**Given dataset: students.csv**

ID,Name,Date,Marks

101,"Asha, R",2026-09-01,80

102,Ravi,01/09/2026,

103,Neha,2026-09-03,NA

**Given code:**

```python
import pandas as pd

df = pd.read_csv("students.csv")

print(df)

print(df.dtypes)
```


In [107]:
import pandas as pd

df = pd.read_csv(r"C:\Users\Lenovo\AppData\Local\Packages\5319275A.WhatsAppDesktop_cv1g1gvanyjgm\LocalState\sessions\BAF5A55F7DDB6ABC349CF4E84348BBF13E2ED9DE\transfers\2026-40\students.csv") #read_csv() reads the CSV file located at the specified path and creates a DataFrame from it.
print(df)
print(df.dtypes)

    ID     Name        Date  Marks
0  101  Asha, R  2026-09-01   80.0
1  102     Ravi  01/09/2026    NaN
2  103     Neha  2026-09-03    NaN
ID         int64
Name         str
Date         str
Marks    float64
dtype: object


### Answers

**2a. Predict how the quoted comma in "Asha, R" will be handled when the CSV file is read.**

**Ans: Quoted comma handling:**

The comma inside `"Asha, R"` is part of the Name field because the value is enclosed in double quotes.

Therefore, Pandas will not treat that comma as a column separator.

---
**2b. Identify the potential issues with the Date and Marks columns.**

**Ans: Potential issues with Date and Marks columns:**

- The `Date` column contains different date formats:
  - `2026-09-01`
  - `01/09/2026`
  - `2026-09-03`
- The `Date` column should be converted into a proper date-time format before performing date-related operations.
- The `Marks` column contains a missing value in row 102.
- The value `NA` in row 103 may represent a missing value, but its meaning should be investigated.

---



**2c. Rewrite the code to handle missing marks and convert the date column into a suitable date-time format.**

**Ans: Code to handle missing marks and convert dates:**


In [108]:
#3. Rewrite the code to handle missing marks and convert the date column into a suitable date-time format.

#**Ans: Code to handle missing marks and convert dates:**

import pandas as pd

df = pd.read_csv(r"C:\Users\Lenovo\AppData\Local\Packages\5319275A.WhatsAppDesktop_cv1g1gvanyjgm\LocalState\sessions\BAF5A55F7DDB6ABC349CF4E84348BBF13E2ED9DE\transfers\2026-40\students.csv") 

df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce") # Convert Marks to numeric, set invalid values to NaN

df["Date"] = pd.to_datetime(   # Convert Date to datetime, set invalid values to NaN
    df["Date"],
    format="mixed",
    dayfirst=True,
    errors="coerce"
)

print(df)
print(df.dtypes)

    ID     Name       Date  Marks
0  101  Asha, R 2026-01-09   80.0
1  102     Ravi 2026-09-01    NaN
2  103     Neha 2026-03-09    NaN
ID                int64
Name                str
Date     datetime64[us]
Marks           float64
dtype: object


**2d. Explain why the value "NA" should be investigated before deciding whether it represents a missing value**

**Ans:** "NA" should be investigated because it may have different meanings, such as Not Available, Not Applicable, or Missing Value. We need to understand what "NA" means in the dataset before treating it as a missing value and cleaning or analyzing the data.

---
## Q3. Salary Outlier and Statistical Measures

### Question

3. **salary column contains the values 25000, 30000, 35000, and 10000000. Explain how the extreme value affects the mean compared with the median. Identify which statistic may better represent a typical employee's salary and justify your choice.**

**Ans:**
The extreme value 10,000,000 greatly increases the mean because the mean uses every value in the calculation.

Mean = (25000 + 30000 + 35000 + 10000000) / 4 = 2,522,500
Median = (30000 + 35000) / 2 = 32,500

The median may better represent a typical employee's salary because it is less affected by the extreme value. The mean is much higher because of the 10,000,000 salary.

---

**3a.Predict the mean and median salary.**


**Ans: Mean and Median Salary:**

Mean salary = 2,522,500

Median salary = 32,500

---

**3b.Identify the effect of the extreme salary on the mean.**

**Ans: Effect of the extreme salary:**

The salary `10,000,000` is an extreme value compared with the other salaries.

It increases the mean significantly because the mean is sensitive to extreme values.

The median is much less affected by the extreme value.

---

**3c.Which statistic may better represent a typical employee's salary in this dataset? Justify your answer.**

**Ans: Statistic that may better represent a typical salary:**

The **median** may better represent a typical employee's salary in this dataset because it is less affected by the extreme salary of `10,000,000`.

The median is `32,500`, which is closer to the salaries of employees A, B, and C, while the mean is `2,522,500` because of the extreme value.

---
**3d.Write code to identify the employee record with the highest salary and explain why the record should be investigated before removing it.**

**Ans: Code to identify the employee with the highest salary:**

```python
highest_salary = df.loc[df["Salary"].idxmax()]

print(highest_salary)


In [109]:
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"], #employee names
    "Salary": [25000, 30000, 35000, 10000000] # salaries of employees, with one outlier
})

print("Mean:", df["Salary"].mean())   #mean the average salary, which can be skewed by the outlier
print("Median:", df["Salary"].median()) #median the middle value of the salaries, which is less affected by the outlier

Mean: 2522500.0
Median: 32500.0


In [110]:
#3d.Code to identify the employee with the highest salary:**

highest_salary = df.loc[df["Salary"].idxmax()]

print(highest_salary)


Employee           D
Salary      10000000
Name: 3, dtype: object


# Section D – Scenario-Based Problems
Covers the application of Pandas concepts to real-world AI/ML scenarios. 

##  Cleaning a Dirty Student Assessment Dataset

Q1.A student assessment dataset contains missing marks, duplicate student records, inconsistent department names such as IT, it, and I.T., and marks stored as text. Develop a cleaning sequence to standardize the data, handle missing values, remove duplicates, and convert marks into a suitable numeric type.

State one risk of deleting missing values without investigating why they are missing.

**Dataset:** `student_assessment_dirty.csv`

**Ans:**

**Risk:** Deleting missing values without investigating why they are missing can remove valid student records and make the analysis biased or misleading.


In [111]:
import pandas as pd

df = pd.read_csv(r"C:\Users\Lenovo\AppData\Local\Packages\5319275A.WhatsAppDesktop_cv1g1gvanyjgm\LocalState\sessions\BAF5A55F7DDB6ABC349CF4E84348BBF13E2ED9DE\transfers\2026-40\student_assessment_dirty.csv")

print(df)                #prints the DataFrame to the console

print(df.dtypes)          #prints the data types of each column in the DataFrame  

   Student_ID Student_Name Department Assessment     Marks
0        S101         Asha         IT     Python        85
1        S102         Ravi         it     Python        78
2        S103         Neha       I.T.     Python        92
3        S104        Karan         HR     Python       NaN
4        S105        Meera         HR     Python    Absent
5        S106        Arjun    Finance     Python         0
6        S107        Priya    finance     Python        67
7        S108        Rahul         IT     Python        75
8        S102         Ravi         it     Python        78
9        S109        Sneha       I.T.     Python       NaN
10       S110        Varun         HR     Python  88 marks
11       S111         Diya    Finance     Python        91
12       S112         Aman         IT     Python       NaN
13       S113        Nisha         HR     Python        64
14       S114        Kabir    Finance     Python       105
Student_ID      str
Student_Name    str
Department      

In [112]:
df

,Student_ID,Student_Name,Department,Assessment,Marks
0,S101,Asha,IT,Python,85
1,S102,Ravi,it,Python,78
2,S103,Neha,I.T.,Python,92
3,S104,Karan,HR,Python,NaN
4,S105,Meera,HR,Python,Absent
5,S106,Arjun,Finance,Python,0
6,S107,Priya,finance,Python,67
7,S108,Rahul,IT,Python,75
8,S102,Ravi,it,Python,78
9,S109,Sneha,I.T.,Python,NaN



# Using Pandas, perform the following cleaning operations in sequence:

**D1. Load the CSV file and inspect the dataset using head(), info(), and isnull().sum()**.

**Ans:** Load the CSV file using pd.read_csv() and inspect it using head(), info(), and isnull().sum().
This helps us understand the data, data types, and missing values before cleaning.


In [113]:
import pandas as pd

df = pd.read_csv(r"C:\Users\Lenovo\AppData\Local\Packages\5319275A.WhatsAppDesktop_cv1g1gvanyjgm\LocalState\sessions\BAF5A55F7DDB6ABC349CF4E84348BBF13E2ED9DE\transfers\2026-40\student_assessment_dirty.csv")

print(df.head())
print(df.info())
print(df.isnull().sum())

  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python     NaN
4       S105        Meera         HR     Python  Absent
<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         12 non-null     str  
dtypes: str(5)
memory usage: 1.0 KB
None
Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           3
dtype: int64


Explanation: The code first loads the student_assessment_dirty.csv file into a Pandas DataFrame,Then:

head() shows the first five records so we can understand the data.

info() gives information about the columns and their data types.

isnull().sum() tells us how many missing values are present in each column.

---

**D2. Standardize department names so that IT, it, and I.T. are represented consistently. Apply the same approach to Finance.**



In [114]:
df["Department"]=(df["Department"].str.strip().str.upper().str.replace(".","",regex=False))
display(df["Department"])

0          IT
1          IT
2          IT
3          HR
4          HR
5     FINANCE
6     FINANCE
7          IT
8          IT
9          IT
10         HR
11    FINANCE
12         IT
13         HR
14    FINANCE
Name: Department, dtype: str

**Explanation:**

The Department column is standardized by removing extra spaces, converting all department names to uppercase, and removing dots. This makes different spellings such as IT, it, and I.T. consistent.

---

**D3.Identify and remove exact duplicate student assessment records.**




In [115]:
before_duplicates = len(df)
df = df.drop_duplicates().copy()
display(df)
display(df.shape)

,Student_ID,Student_Name,Department,Assessment,Marks
0,S101,Asha,IT,Python,85
1,S102,Ravi,IT,Python,78
2,S103,Neha,IT,Python,92
3,S104,Karan,HR,Python,NaN
4,S105,Meera,HR,Python,Absent
5,S106,Arjun,FINANCE,Python,0
6,S107,Priya,FINANCE,Python,67
7,S108,Rahul,IT,Python,75
9,S109,Sneha,IT,Python,NaN
10,S110,Varun,HR,Python,88 marks


(14, 5)

**Explination:** The drop_duplicates() method removes exact duplicate student assessment records so that the same record is not counted more than once.

---
**D4.Convert the Marks column into a numeric format, handling "88 marks", "Absent", "NA", and blank entries appropriately.**


In [116]:
# Before converting Marks to numeric, identify the reason for missing marks
import numpy as np
df["missing_marks"] = np.where(
    df["Marks"] == "Absent",
    "Absent",
    np.where(
        df["Marks"].isna(),"not_entered","Marks"))
#cleaning
df["Marks"] = (
    df["Marks"]
    .astype("string")
    .str.strip()
    .replace({
        "Absent": pd.NA,
        "NA": pd.NA,
        "": pd.NA
    })
    .str.replace(" marks", "", regex=False)
)

df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")
display(df["Marks"])

0       85
1       78
2       92
3     <NA>
4     <NA>
5        0
6       67
7       75
9     <NA>
10      88
11      91
12    <NA>
13      64
14     105
Name: Marks, dtype: Int64

**Explanation:** The Marks column is cleaned by removing extra spaces and text such as " marks", handling "Absent", "NA", and blank entries as missing values, and finally converting the column into a numeric format.

---
**D5.Investigate missing marks and distinguish absent students from students whose marks have not been entered. State one risk of deleting missing values without investigating why they are missing.**



In [117]:
missing_marks = df[df["Marks"].isna()]

display(missing_marks[
    ["Student_ID", "Student_Name", "Department", "Marks"]
])

print("\nMissing marks:", df["Marks"].isna().sum())

,Student_ID,Student_Name,Department,Marks
3,S104,Karan,HR,<NA>
4,S105,Meera,HR,<NA>
9,S109,Sneha,IT,<NA>
12,S112,Aman,IT,<NA>



Missing marks: 4


**Explination:** Missing marks should be investigated before deleting the corresponding records. For example, Absent may represent a student who did not attend the assessment, while a blank value may mean that the mark was not entered.

**Risk:** Risk of deleting missing values
Deleting missing records without investigating why they are missing can remove useful information and may introduce bias into the analysis.

---
**D6.Identify marks outside the valid range of 0–100 and decide how they should be handled.**



In [118]:

# 6. Identify invalid marks outside 0–100
invalid_mask = (df["Marks"] < 0) | (df["Marks"] > 100)

invalid_marks = df[invalid_mask]

print("\nInvalid marks:")
print(invalid_marks)




Invalid marks:
   Student_ID Student_Name Department Assessment  Marks missing_marks
14       S114        Kabir    FINANCE     Python    105         Marks


**Explination:** Display the cleaned DataFrame and report how many records were removed, missing marks remain, and invalid marks were found.
This gives a clear summary of the cleaning process and its results.

---
**D7.Display the cleaned DataFrame and report the number of records removed, missing marks remaining, and invalid marks identified.**

In [119]:
print(df)
records_removed = before_duplicates - len(df)
print("Records removed:", records_removed)
print("Missing marks remaining:", df["Marks"].isna().sum())
print("Invalid marks identified:", len(invalid_marks))

   Student_ID Student_Name Department Assessment  Marks missing_marks
0        S101         Asha         IT     Python     85         Marks
1        S102         Ravi         IT     Python     78         Marks
2        S103         Neha         IT     Python     92         Marks
3        S104        Karan         HR     Python   <NA>   not_entered
4        S105        Meera         HR     Python   <NA>        Absent
5        S106        Arjun    FINANCE     Python      0         Marks
6        S107        Priya    FINANCE     Python     67         Marks
7        S108        Rahul         IT     Python     75         Marks
9        S109        Sneha         IT     Python   <NA>   not_entered
10       S110        Varun         HR     Python     88         Marks
11       S111         Diya    FINANCE     Python     91         Marks
12       S112         Aman         IT     Python   <NA>   not_entered
13       S113        Nisha         HR     Python     64         Marks
14       S114       

**Explination:** displaying the final cleaned dataset and summarizes how many duplicate records were removed, how many marks are still missing, and how many invalid marks were identified.



## My Experience

Working on this Pandas data-cleaning assignment helped me understand how real-world datasets can contain missing values, duplicate records, inconsistent text, and incorrect data types.

I learned how to inspect a dataset, standardize department names, remove duplicate records, clean and convert marks into numeric values, and identify invalid marks.

One important thing I learned is that missing or incorrect data should not be deleted immediately. We should first understand why the data is missing or invalid and then decide how to handle it.

Overall, this assignment gave me a better understanding of the importance of data cleaning before performing data analysis.

